File: algoproInPython/three_sum.py
"""
@Author:    Hai Nguyen
@Date:      Nov 7th, 2019
@Source:    AlgoPro
@Task:      Determine a binary tree is a valid binary search tree.
@Desc:      
            
"""


class Solution:
    def threeSumBruteForce(self, nums):
        res = set()
        nums.sort()
        for i in range(len(nums) - 2):
            for j in range(i + 1, len(nums) - 1):
                for k in range(j + 1, len(nums)):
                    if (nums[i] + nums[j] + nums[k] == 0):
                        res.add((nums[i], nums[j], nums[k]))
        return list(res)

    def threeSum(self, nums):
        res = []
        nums.sort()
        for i in range(len(nums) - 2):
            if (nums[i] == nums[i - 1] and i > 0):
                continue
            #two sum
            j = i + 1
            k = len(nums) - 1
            while (j < k):
                if (nums[i] + nums[j] + nums[k] == 0):
                    res.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while j < k and nums[k] == nums[k - 1]:
                        k -= 1
                    j += 1
                    k -= 1
                elif nums[i] + nums[j] + nums[k] > 0:
                    k -= 1
                else:
                    j += 1
            #end of two sum
        return res


print(Solution().threeSumBruteForce([-1, 0, 1, 2, -1, -4]))
print(Solution().threeSum([-1, 0, 1, 2, -1, -4]))

File: algoproInJava/src/com/algopro/ThreeSum/ThreeSum.java
package com.algopro.ThreeSum;

import java.util.ArrayList;
import java.util.Arrays;

public class ThreeSum {

	public static int[][] Solution(int[] nums) {
		ArrayList<Integer[]> list = new ArrayList<>();
		Arrays.sort(nums);
		//
		for (int i = 0; i < nums.length - 2; i++) {
			if (i > 0 && nums[i] == nums[i - 1])
				continue;
			// two sum
			int j = i + 1;
			int k = nums.length - 1;
			while (j < k) {
				int sum = nums[i] + nums[j] + nums[k];
				if (sum == 0) {
					list.add(new Integer[] { nums[i], nums[j], nums[k] });
					while (j < k && nums[j] == nums[j + 1])
						j++;
					while (j < k && nums[k] == nums[k - 1])
						k--;
					j++;
					k--;
				} else if (sum > 0) {
					k--;
				} else {
					j++;
				}
			}
			// end of two sum
		}
		//
		int[][] res = new int[list.size()][3];
		for (int i = 0; i < list.size(); i++) {
			Integer[] tmp = list.get(i);
			res[i] = new int[] { tmp[0], tmp[1], tmp[2] };
		}
		return res;
	}

	public static void main(String[] args) {
		int[] test = { -1, 0, 1, 2, -1, -4 };
		int[][] solutions = Solution(test);
		for (int i = 0; i < solutions.length; i++) {
			for (int j = 0; j < solutions[i].length; j++) {
				System.out.print(solutions[i][j] + " ");
			}
			System.out.println();
		}
	}

}
